File: workers/image_extractor/services/extractor_service.py
import json
import logging
import uuid

import httpx

from shared.clients.s3_client import S3Client
from shared.clients.sqs_client import SQSClient
from workers.image_extractor.services.explainer_factory import ExplainerFactory

logger = logging.getLogger(__name__)


class ExtractorService:
    # pylint: disable=too-few-public-methods
    def __init__(
        self,
        sqs_client: SQSClient,
        s3_client: S3Client,
        writer_queue_url: str,
        images_bucket: str,
        llm_provider: str = "openai",
        llm_api_key: str | None = None,
    ):  # pylint: disable=too-many-arguments,too-many-positional-arguments
        self.__sqs_client = sqs_client
        self.__s3_client = s3_client
        self.__writer_queue_url = writer_queue_url
        self.__images_bucket = images_bucket
        self.__llm = ExplainerFactory.get_explainer(llm_provider, llm_api_key)

    async def process_message(self, message_body: str) -> None:
        try:
            body = json.loads(message_body)
            image_url = body.get("url")
            original_url = body.get("original_url")
            scraping_id = body.get("scraping_id")

            if not image_url:
                logger.warning("No image URL in message")
                return

            if not scraping_id:
                logger.warning("No scraping_id in message")
                return

            logger.info("Processing image: %s for scraping %s", image_url, scraping_id)

            # 1. Download image
            s3_path = None
            try:
                async with httpx.AsyncClient() as client:
                    resp = await client.get(image_url, timeout=10.0)
                    if resp.status_code == 200:
                        # Upload to S3
                        s3_path = await self.__upload_image_to_s3(
                            resp.content, image_url, scraping_id
                        )
            except Exception as e:  # pylint: disable=broad-exception-caught
                logger.error("Failed to handle image persistence to S3: %s", e)

            # 3. Get Explanation via LangChain
            explanation = ExplainerFactory.explain_image(self.__llm, image_url)
            logger.info("Generated explanation for %s", image_url)

            # Send to Writer
            writer_msg = {
                "type": "image_explanation",
                "url": image_url,
                "original_url": original_url,
                "page_url": original_url,
                "scraping_id": scraping_id,
                "explanation": explanation,
                "s3_path": s3_path,
            }

            await self.__sqs_client.send_message(writer_msg, self.__writer_queue_url)
            logger.info("Sent extraction info for %s to writer queue", image_url)

        except Exception as e:  # pylint: disable=broad-exception-caught
            # Catch-all to prevent worker crash on single message failure
            logger.error("Error processing image extraction: %s", e)

    async def __upload_image_to_s3(
        self, content: bytes, image_url: str, scraping_id: int
    ) -> str | None:
        try:
            ext = image_url.split(".")[-1].split("?")[0] or "bin"
            s3_key = f"{scraping_id}/{uuid.uuid4()}.{ext}"
            s3_path = await self.__s3_client.upload_bytes(
                content, self.__images_bucket, s3_key, "image/jpeg"
            )
            logger.info("Uploaded image to %s", s3_path)
            return s3_path
        except Exception as e:  # pylint: disable=broad-exception-caught
            logger.error("Internal S3 upload failure: %s", e)
            return None

File: workers/image_extractor/services/extractor_service.go
package services

import (
	"context"
	"fmt"
	"log"
	"mime"
	"path/filepath"
	"strings"

	"workers/image_extractor/domain"
	"workers/image_extractor/repositories"

	"github.com/google/uuid"
)

type SQSRepository interface {
	SendMessage(ctx context.Context, queueURL string, body interface{}) error
}

type S3Repository interface {
	UploadBytes(ctx context.Context, bucket, key string, data []byte, contentType string) (string, error)
}

type HTTPRepository interface {
	DownloadImage(ctx context.Context, url string) ([]byte, string, error)
}

type ExtractorService struct {
	sqsRepo                SQSRepository
	s3Repo                 S3Repository
	httpRepo               HTTPRepository
	writerQueueURL         string
	imageExplainerQueueURL string
	imagesBucket           string
	imageExplainerEnabled  bool
	otelClient             repositories.TelemetryClient
}

func NewExtractorService(
	sqsRepo SQSRepository,
	s3Repo S3Repository,
	httpRepo HTTPRepository,
	writerQueueURL string,
	imageExplainerQueueURL string,
	imagesBucket string,
	imageExplainerEnabled bool,
	otelClient repositories.TelemetryClient,
) *ExtractorService {
	if otelClient == nil {
		otelClient = repositories.NewNoopTelemetryClient()
	}
	return &ExtractorService{
		sqsRepo:                sqsRepo,
		s3Repo:                 s3Repo,
		httpRepo:               httpRepo,
		writerQueueURL:         writerQueueURL,
		imageExplainerQueueURL: imageExplainerQueueURL,
		imagesBucket:           imagesBucket,
		imageExplainerEnabled:  imageExplainerEnabled,
		otelClient:             otelClient,
	}
}

func (s *ExtractorService) ProcessMessage(ctx context.Context, msg domain.ImageMessage) error {
	ctx, span := s.otelClient.StartSpan(ctx, "ExtractorService.ProcessMessage",
		repositories.WithAttribute("url", msg.URL),
		repositories.WithAttribute("scrapingID", msg.ScrapingID),
	)
	defer span.End()

	log.Printf("Processing image: %s for scraping %d (AI Enabled: %v)", msg.URL, msg.ScrapingID, s.imageExplainerEnabled)

	// 1. Download image
	data, contentType, err := s.httpRepo.DownloadImage(ctx, msg.URL)
	var s3Path string
	if err != nil {
		log.Printf("Failed to download image %s: %v", msg.URL, err)
	} else {
		// 2. Upload to S3
		ext := s.getExtension(msg.URL, contentType)
		s3Key := fmt.Sprintf("%d/%s.%s", msg.ScrapingID, uuid.New().String(), ext)

		path, err := s.s3Repo.UploadBytes(ctx, s.imagesBucket, s3Key, data, contentType)
		if err != nil {
			log.Printf("Internal S3 upload failure for %s: %v", msg.URL, err)
		} else {
			s3Path = path
			log.Printf("Uploaded image to %s", s3Path)
		}
	}

	// 3. Send to Writer (Metadata)
	writerMsg := domain.WriterMessage{
		Type:        "image_explanation",
		URL:         msg.URL,
		OriginalURL: msg.OriginalURL,
		PageURL:     msg.OriginalURL,
		ScrapingID:  msg.ScrapingID,
		S3Path:      s3Path,
	}
	if err := s.sqsRepo.SendMessage(ctx, s.writerQueueURL, writerMsg); err != nil {
		span.RecordError(err)
		span.SetStatus("error", err.Error())
		return fmt.Errorf("failed to send image metadata to writer: %w", err)
	}
	log.Printf("Sent image metadata for %s to writer queue", msg.URL)

	// 4. Send to Explainer (if S3 upload succeeded AND explainer is enabled)
	if s3Path != "" && s.imageExplainerEnabled {
		explainerMsg := domain.ImageExtractorMessage{
			ImageURL:    msg.URL,
			OriginalURL: msg.OriginalURL,
			ScrapingID:  msg.ScrapingID,
			S3Path:      s3Path,
		}
		if err := s.sqsRepo.SendMessage(ctx, s.imageExplainerQueueURL, explainerMsg); err != nil {
			log.Printf("failed to send image to explainer queue: %v", err)
		} else {
			log.Printf("Sent image to explainer queue: %s", msg.URL)
		}
	} else if s3Path != "" {
		log.Printf("Skipping image explanation for %s (disabled globally)", msg.URL)
	}

	span.SetStatus("ok", "success")
	return nil
}

func (s *ExtractorService) getExtension(url, contentType string) string {
	ext := "bin"

	// Try from content type
	if contentType != "" {
		exts, err := mime.ExtensionsByType(contentType)
		if err == nil && len(exts) > 0 {
			ext = strings.TrimPrefix(exts[0], ".")
		}
	}

	// Fallback to URL extension if still bin or no content type
	if ext == "bin" || contentType == "" {
		urlExt := filepath.Ext(strings.Split(url, "?")[0])
		if urlExt != "" && len(urlExt) < 6 {
			ext = strings.TrimPrefix(urlExt, ".")
		}
	}

	return ext
}
